src/metabase/csv.clj

```diff
@@ -63,11 +62,7 @@
     (nil? type-b)        type-a
     (isa? type-a type-b) type-b
     (isa? type-b type-a) type-a
-    :else
-    (let [common-ancestors (set/intersection (ancestors type-a) (ancestors type-b))]
-      (if (seq common-ancestors)
-        (first common-ancestors)
-        (throw (Exception. (trs "Unexpected type combination in the same column: {0} and {1}" type-a type-b)))))))
+    :else (throw (Exception. (trs "Unexpected type combination in the same column: {0} and {1}" type-a type-b)))))
```

Comment:
instead of this maybe return `::ununifiable` or `{:unreconcilable [::boolean ::datetime]}` or something and then we can get all of the type errors at the end rather than making people fix the csv one row at a time and we keep yelling at them.

But feel free to delay that for later and we can merge this PR right now.

Reply:
bonus points if we can keep the column names and a representative value around.
"Can't determine type of columns:
 created-at: "2023-05-16" (date), "bark" (string)
is-valid:  "true" (boolean), 3.23 (integer)
" 
or something like that

Reply:
Yeah, I've been wondering about this. The current agreement is that any possible type mismatch will be resolved by walking up the chain until you get to a more permissive type, and AFAIK there's nothing you can put in a CSV that doesn't work as a SQL `text` column. Which means that we'll never throw a type mismatch error, even once the type hierarchy goes from being a list to a proper tree (with datetimes).

This means that the `:else` clause in its entirety is defensive programming and potentially should be left out. If you think it should that's fine. It's in right now to make development easier (concrete example: when I screwed up the type hierarchy and got a useful error message in tests instead of a null pointer exception), but maybe that's not a good reason to keep dead code in.